### PR TITLE
Field Level Encryption Support.

### LINKED ISSUE #763: Field Level Encryption [DATACOUCH-455]
**[Domenico Ursoleo](https://jira.spring.io/secure/ViewProfile.jspa?name=udomenico)** opened **[DATACOUCH-455](https://jira.spring.io/browse/DATACOUCH-455?redirect=false)** and commented

Couchbase has a Field Level Encryption feature that encrypt field of a Document simply putting the `@EncryptField` annotation on a field of the document model and addind a CryptoManager to the environment,

With this configuration:

```java
@Configuration
@EnableCouchbaseRepositories(basePackages = { "com.example.demo.db" })
public class MyCouchbaseConfig extends AbstractCouchbaseConfiguration {
 
@Override 
protected CouchbaseEnvironment getEnvironment() { 
try {
 JceksKeyStoreProvider kp = new JceksKeyStoreProvider("secret");
 kp.publicKeyName("mypublickey");
 kp.storeKey("mypublickey",
           "!mysecretkey#9^5usdk39d&dlf)03sL".getBytes(Charset.forName("UTF-8")));
 kp.signingKeyName("HMACsecret");
 kp.storeKey("HMACsecret", "myauthpassword".getBytes(Charset.forName("UTF-8")));
 AES256CryptoProvider aes256CryptoProvider = new AES256CryptoProvider(kp);
 CryptoManager cryptoManager = new CryptoManager();
 cryptoManager.registerProvider("MyAESProvider", aes256CryptoProvider);
 return 
    DefaultCouchbaseEnvironment.builder().cryptoManager(cryptoManager).build();
 } catch (Exception ex) { return null; } }
}
```

 Given the model

 

```java
@Document
public class User {
@Id
@GeneratedValue(strategy = GenerationStrategy.UNIQUE)
private String id;

@Field("user")
private String userName;

@EncryptedField(provider = "MyAESProvider") 
@Field("pwd") 
private String password;

// Setters & getters
}
```

Persisting it via a simple CrudRepository should encrypt the password like calling the native couchbase java client:

```java
EntityDocument<User> document = EntityDocument.create(user);          bucket.repository().upsert(document);
 
```

 

 

---

**Affects:** 3.1.8 (Lovelace SR8)

1 votes, 2 watchers

### PR DESCRIPTION
**Pull request by mikereiche**
Note that an @Encrypted property object that contains an @Encrypted property is
not supported and will result in an exception.

Closes #763.

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATACOUCH).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: spring-data-couchbase/src/main/java/org/springframework/data/couchbase/config/AbstractCouchbaseConfiguration.java [hunk 1]
```diff
@@ -379,11 +400,28 @@ protected boolean autoIndexCreation() {
...
 	 * default.
 	 *
+	 * @param cryptoManagerOptional optional cryptoManager. Make varargs for backwards compatibility.
 	 * @return must not be {@literal null}.
 	 */
 	@Bean(name = BeanNames.COUCHBASE_CUSTOM_CONVERSIONS)
-	public CustomConversions customConversions() {
-		return new CouchbaseCustomConversions(Collections.emptyList());
+	public CustomConversions customConversions(CryptoManager... cryptoManagerOptional) {
+		assert (cryptoManagerOptional == null || cryptoManagerOptional.length <= 1);
```
Comment (dnault) [inline spring-data-couchbase/src/main/java/org/springframework/data/couchbase/config/AbstractCouchbaseConfiguration.java:408]: What do you think of these alternatives stolen from StackOverflow?

```
public CustomConversions customConversions(
    Optional<CryptoManager> cryptoManagerOptional
)
```

Or this:

```
public CustomConversions customConversions(
    @Autowired(required = false) CryptoManager cryptoManager
)
```

Reference: https://stackoverflow.com/questions/46704950/how-to-define-optional-parameters-dependencies-in-the-bean-method-in-spring

Comment (mikereiche) [reply spring-data-couchbase/src/main/java/org/springframework/data/couchbase/config/AbstractCouchbaseConfiguration.java:408]: The first one changes a public API so is not possible.
The second one - cryptoManager is not a bean and therefore cannot be instantiated with @Autowired. I also have doubts about using @Autowired in the configuration class.  And also bean-ifying things that don't absolutely need to be beans.

Comment (dnault) [reply spring-data-couchbase/src/main/java/org/springframework/data/couchbase/config/AbstractCouchbaseConfiguration.java:408]: Changing `foo()` to `foo(T...)` is a source-compatible change, but I suspect it's not binary-compatible. (Maybe binary compatibility is not a concern?)

**Review by dnault — COMMENTED** [review]
Looks like I missed the deadline -- sorry!

**Pull request merged**